pkg/logql/stats/context.go

```diff
@@ -0,0 +1,193 @@
+/*
+Package stats provides primitives for recording metrics across the query path.
+Statistics are passed through the query context.
+To start a new query statistics context use:
+
+	ctx := stats.NewContext(ctx)
+
+Then you can update statistics by mutating data by using:
+
+	stats.GetChunkData(ctx)
+	stats.GetIngesterData(ctx)
+	stats.GetStoreData
+
+Finally to get a snapshot of the current query statistic use
+
+	stats.Snapshot(ctx,time.Since(start))
+
+Ingester statistics are sent across the GRPC stream using Trailers
+see https://github.com/grpc/grpc-go/blob/master/Documentation/grpc-metadata.md
+*/
+package stats
+
+import (
+	"context"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+)
+
+type ctxKeyType string
+
+const (
+	trailersKey ctxKeyType = "trailers"
+	chunksKey   ctxKeyType = "chunks"
+	ingesterKey ctxKeyType = "ingester"
+	storeKey    ctxKeyType = "store"
+)
+
+// Result contains LogQL query statistics.
+type Result struct {
+	Ingester Ingester
+	Store    Store
+	Summary  Summary
+}
+
+// Log logs a query statistics result.
+func Log(log log.Logger, r Result) {
```

Comment:
can you add some comments to these describing what they refer to? I'm having a hard time understanding`Uncompressed vs Decompressed vs Compressed`. I'm guessing one is before decompression, one is after decompression, and one is after recompression?

Reply:
- Uncompressed is data that was processed but not compressed from headchunk.
- Decompressed is data that  was processed from decompression store or ingester.
- Compressed is data that that was decompressed. but the value before decompression.

Reply:
I'm happy to change the names, or document it better where would it be better to document this ?

@slim-bean  ??

Reply:
I think if you put your comment in the code as a comment it'd be great.

Reply:
2 thoughts:

Do we really care that they came out of the head chunk?  Should we just add the line and byte count to the decompressed counts?

If we do care, i would suggest naming them:

```
HeadChunkBytes
HeadChunkLines
```